haptools/sim_genotype.py

```diff
@@ -68,219 +89,167 @@ def output_vcf(breakpoints, chroms, model_file, vcf_file, sampleinfo_file, regio
-        in_vcf = GenotypesPLINK(in_vcf)
+    # TODO Load populations and samples from sample info and breakpoints as 2 matrices with size variants x samples x 2
+    # Use search_sorted() numpy function to find indices in populations/samples in order to determine population and sample info
+    ref_vars = vcf.variants
+    output_gts = np.empty((int(len(breakpoints)//2), len(vcf.variants), 2), dtype=vcf.data.dtype)
+    output_pops = np.empty((int(len(breakpoints)//2), len(vcf.variants), 2), dtype=np.uint8)
+    output_labels = np.empty((int(len(breakpoints)//2), len(vcf.variants), 2), dtype=object)
+
+    # cover "chr" prefix cases
+    if ref_vars["chrom"][0].startswith("chr"):
+        cur_chrom = "chr"
     else:
-        in_vcf = GenotypesRefAlt(in_vcf)
-    in_vcf.read(region=f"{region['chr']}:{region['start']}-{region['end']}")
-    # TODO: check with someone, do we need to do this QC?
-    in_vcf.check_missing(discard_also=True)
-    in_vcf.check_biallelic(discard_also=True)
-    in_vcf.check_phase()
-
-    # initialize output writer
-    gts = GenotypesPLINK(out)
-    gts.samples = out_samples
-    gts.variants = in_vcf.variants
-    gts.data = np.empty((len(out_samples), len(gts.variants), 2), dtype=in_vcf.data.dtype)
-
-    # Now we just fill out gts.data
-    # TODO: figure out if there's a way to optimize the following lines of code so that
-    # this for loop is performed in numpy rather than in python? This is especially
-    # relevant for situations in which the input is PGEN b/c it's probably just as fast
-    # to stream the VCF like this otherwise
-    for var_idx, var in enumerate(gts.variants):
-        # parse chromosome
-        chrom = re.search(r'X|\d+', var["chrom"]).group()
-        if chrom not in chroms: continue
-        if chrom == 'X':
-            chrom = 23
-        for hap in range(len(hapblock_samples)):
-            sample_num = hap // 2
-            # If breakpoint end coord is < current variant update breakpoint
-            bkp = breakpoints[hap][current_bkps[hap]]
-            # Note: for some reason, var.start in _write_vcf() is always equal to var["pos"]-1 in _write_pgen()
-            # We should probably investigate at some point
-            while bkp.get_chrom() < int(chrom) or (bkp.get_chrom() == int(chrom) and bkp.get_end_coord() < int(var["pos"])-1):
-                current_bkps[hap] += 1
-                bkp = breakpoints[hap][current_bkps[hap]]
-            var_sample = hapblock_samples[hap][current_bkps[hap]]
-            if hap % 2 == 0:
-                # store variant
-                if hap > 0:
-                    gts.data[sample_num-1, var_idx] = tuple(gt)
-                gt = []
-            gt.append(int(in_vcf.data[var_sample, var_idx, hap % 2]))
-        sample_num = hap // 2
-        gts.data[sample_num, var_idx] = tuple(gt)
+        cur_chrom = ""
+
+    # create samples x variants x 2 matrix of populations
+    for hap_ind, haplotype in enumerate(breakpoints):
```

Comment:
Nice work, @mlamkin7 ! Just one question:

Is this loop iterating over each sample? If so, this might get slow for large sample sizes.

Reply:
It is however, to replace this loop would require extensive refactoring of breakpoint generation since that also iterates over every sample.